src/common/io_exerciser/RadosIo.cc

```diff
@@ -0,0 +1,279 @@
+#include "RadosIo.h"
+
+#include "DataGenerator.h"
+
+using RadosIo = ceph::io_exerciser::RadosIo;
+
+RadosIo::RadosIo(librados::Rados& rados,
+        boost::asio::io_context& asio,
+        const std::string& pool,
+        const std::string& oid,
+        uint64_t block_size,
+        int seed,
+	int threads,
+        ceph::mutex& lock,
+        ceph::condition_variable& cond) :
+  Model(oid, block_size),
+  rados(rados),
+  asio(asio),
+  om(std::make_unique<ObjectModel>(oid, block_size, seed)),
+  db(data_generation::DataGenerator::create_generator(
+      data_generation::GenerationType::HeaderedSeededRandom, *om)),
+  pool(pool),
+  threads(threads),
+  lock(lock),
+  cond(cond),
+  outstanding_io(0)
+{
+  int rc;
+  rc = rados.ioctx_create(pool.c_str(), io);
+  ceph_assert(rc == 0);
+  allow_ec_overwrites(true);
+}
+
+RadosIo::~RadosIo()
+{
+}
+
+void RadosIo::start_io()
+{
+  std::lock_guard l(lock);
+  outstanding_io++;
+}
+
+void RadosIo::finish_io()
+{
+  std::lock_guard l(lock);
+  ceph_assert(outstanding_io > 0);
+  outstanding_io--;
+  cond.notify_all();
+}
+
+void RadosIo::wait_for_io(int count)
+{
+  std::unique_lock l(lock);
+  while (outstanding_io > count) {
+    cond.wait(l);
+  }
+}
+
+void RadosIo::allow_ec_overwrites(bool allow)
+{
+  int rc;
+  bufferlist inbl, outbl;
+  std::string cmdstr =
+    "{\"prefix\": \"osd pool set\", \"pool\": \"" + pool + "\", \
+      \"var\": \"allow_ec_overwrites\", \"val\": \"" +
+    (allow ? "true" : "false") + "\"}";
+  rc = rados.mon_command(cmdstr, inbl, &outbl, nullptr);
+  ceph_assert(rc == 0);
+}
+
+RadosIo::AsyncOpInfo::AsyncOpInfo(uint64_t offset1, uint64_t length1,
+                                  uint64_t offset2, uint64_t length2,
+                                  uint64_t offset3, uint64_t length3 ) :
+  offset1(offset1), length1(length1),
+  offset2(offset2), length2(length2),
+  offset3(offset3), length3(length3)
+{
+
+}
+
+bool RadosIo::readyForIoOp(IoOp &op)
+{
+  ceph_assert(lock.is_locked_by_me()); //Must be called with lock held
```

Comment:
Line fails build complaining about missing function is_locked_by_me().
Build fails when doing a DEBUG build

Reply:
I have a PR open to fix this here: https://github.com/ceph/ceph/pull/60211, as this code has now unfortunately merged with this line in.

However I noticed the opposite when opening the PR - that the build fails when doing a RELEASE/RelWithDebInfo build and succeeds when doing a DEBUG build. Can I confirm the build type you are doing please? If it is debug, are you able to provide any more information?

Reply:
I was doing DEBUG build (which activates the assert code) and failed with a complain about  non-existing member function is_locked_by_me() for BOOST::lock.

I comment out the assert line in my code and now it builds fine